File: planning_pipeline/pipeline.py
"""Main planning pipeline with interactive checkpoints."""

from pathlib import Path
from datetime import datetime
from typing import Any, Optional

from .beads_controller import BeadsController
from .steps import step_research, step_planning, step_phase_decomposition, step_beads_integration, step_memory_sync
from .checkpoints import interactive_checkpoint_research, interactive_checkpoint_plan
from .step_decomposition import step_requirement_decomposition


class PlanningPipeline:
    """Interactive planning pipeline with deterministic control.

    Orchestrates 6 steps:
    1. Research - Analyze codebase and create research document
    2. Memory Sync - Record research and clear context
    3. Requirement Decomposition - Generate structured requirements hierarchy
    4. Planning - Create implementation plan from research
    5. Phase Decomposition - Split plan into phase files
    6. Beads Integration - Create issues and dependencies
    """

    def __init__(self, project_path: Path):
        self.project_path = Path(project_path).resolve()
        self.beads = BeadsController(project_path)

    def run(
        self,
        research_prompt: str,
        ticket_id: Optional[str] = None,
        auto_approve: bool = False
    ) -> dict[str, Any]:
        """Run the complete planning pipeline.

        Args:
            research_prompt: The initial research question or topic
            ticket_id: Optional ticket ID for tracking
            auto_approve: If True, skip interactive checkpoints

        Returns:
            Dictionary with keys:
            - success: bool
            - steps: dict with results from each step
            - plan_dir: path to plan directory
            - epic_id: ID of created epic
        """
        results = {
            "started": datetime.now().isoformat(),
            "ticket_id": ticket_id,
            "steps": {}
        }

        # Step 1: Research (may loop on revise/restart)
        current_prompt = research_prompt
        additional_context = ""

        while True:
            print("\n" + "="*60)
            print("STEP 1/6: RESEARCH PHASE")
            print("="*60)

            research = step_research(self.project_path, current_prompt)
            results["steps"]["research"] = research

            if not research["success"]:
                from .checkpoint_manager import write_checkpoint
                write_checkpoint(
                    self.project_path,
                    "research-failed",
                    [],
                    [research.get("error", "Unknown error")]
                )
                results["success"] = False
                results["failed_at"] = "research"
                return results

            if auto_approve:
                break

            checkpoint = interactive_checkpoint_research(research)
            action = checkpoint.get("action", "continue")

            if action == "continue":
                additional_context = "\n".join(checkpoint.get("answers", []))
                break
            elif action == "revise":
                # Add revision context to prompt and re-run
                revision = checkpoint.get("revision_context", "")
                if revision:
                    current_prompt = f"{current_prompt}\n\nAdditional context:\n{revision}"
                print("\nRe-running research with additional context...")
            elif action == "restart":
                # Collect new prompt
                print("\nEnter new research prompt (empty line to finish):")
                lines = []
                while True:
                    line = input()
                    if not line:
                        break
                    lines.append(line)
                new_prompt = "\n".join(lines)
                if new_prompt.strip():
                    current_prompt = new_prompt
                print("\nRestarting research with new prompt...")
            else:  # action == "exit"
                results["success"] = False
                results["stopped_at"] = "research"
                return results

        # Memory sync between research and planning
        print("\n" + "="*60)
        print("MEMORY SYNC: Recording research & clearing context")
        print("="*60)

        session_id = f"research-{datetime.now().strftime('%Y%m%d-%H%M%S')}"
        memory_result = step_memory_sync(
            self.project_path,
            research["research_path"],
            session_id
        )
        results["steps"]["memory_sync"] = memory_result

        if memory_result.get("episode_recorded"):
            print("  ✓ Episodic memory recorded")
        if memory_result.get("context_compiled"):
            print("  ✓ Working context compiled")
        if memory_result.get("context_cleared"):
            print("  ✓ Claude context cleared")

        # Step 2: Requirement Decomposition (with retry loop)
        while True:
            print("\n" + "="*60)
            print("STEP 2/6: REQUIREMENT DECOMPOSITION")
            print("="*60)

            req_decomp = step_requirement_decomposition(
                self.project_path,
                research["research_path"]
            )
            results["steps"]["requirement_decomposition"] = req_decomp

            if req_decomp["success"]:
                print(f"\nDecomposed into {req_decomp['requirement_count']} requirements")
                print(f"Hierarchy: {req_decomp['hierarchy_path']}")
                print(f"Diagram: {req_decomp['diagram_path']}")
                break

            # Decomposition failed
            print(f"\nDecomposition failed: {req_decomp.get('error', 'Unknown error')}")

            if auto_approve:
                print("Auto-approve mode: skipping to planning")
                break

            # Interactive prompt
            print("\nOptions:")
            print("  (R)etry - Try decomposition again")
            print("  (C)ontinue - Skip decomposition and proceed to planning")

            choice = input("\nChoice [R/c]: ").strip().lower()
            if choice == 'r' or choice == '':
                print("\nRetrying decomposition...")
                continue
            else:
                print("\nSkipping decomposition, continuing to planning...")
                break

        # Step 3: Planning (may loop on feedback)
        while True:
            print("\n" + "="*60)
            print("STEP 3/6: PLANNING PHASE")
            print("="*60)

            planning = step_planning(
                self.project_path,
                research["research_path"],
                additional_context
            )
            results["steps"]["planning"] = planning

            if not planning["success"]:
                from .checkpoint_manager import write_checkpoint
                write_checkpoint(
                    self.project_path,
                    "planning-failed",
                    [str(Path(research["research_path"]).resolve())],
                    [planning.get("error", "Unknown error")]
                )
                results["success"] = False
                results["failed_at"] = "planning"
                return results

            if auto_approve:
                break
            else:
                plan_checkpoint = interactive_checkpoint_plan(planning)
                if plan_checkpoint["continue"]:
                    break
                additional_context = plan_checkpoint["feedback"]
                print("\nRe-running planning with feedback...")

        # Step 4: Phase Decomposition
        print("\n" + "="*60)
        print("STEP 4/6: PHASE DECOMPOSITION")
        print("="*60)

        if not planning.get("plan_path"):
            print("ERROR: Planning completed but no plan file path was extracted.")
            print("Check Claude output for the expected plan file location.")
            results["success"] = False
            results["failed_at"] = "phase_decomposition"
            results["error"] = "No plan_path extracted from planning step"
            return results

        decomposition = step_phase_decomposition(
            self.project_path,
            planning["plan_path"]
        )
        results["steps"]["decomposition"] = decomposition

        if not decomposition["success"]:
            from .checkpoint_manager import write_checkpoint
            write_checkpoint(
                self.project_path,
                "decomposition-failed",
                [
                    str(Path(research["research_path"]).resolve()),
                    str(Path(planning["plan_path"]).resolve())
                ],
                [decomposition.get("error", "Unknown error")]
            )
            results["success"] = False
            results["failed_at"] = "decomposition"
            return results

        print(f"\nCreated {len(decomposition['phase_files'])} phase files")

        # Step 5: Beads Integration
        print("\n" + "="*60)
        print("STEP 5/6: BEADS INTEGRATION")
        print("="*60)

        epic_title = f"Plan: {ticket_id}" if ticket_id else f"Plan: {datetime.now().strftime('%Y-%m-%d')}"
        beads = step_beads_integration(
            self.project_path,
            decomposition["phase_files"],
            epic_title
        )
        results["steps"]["beads"] = beads

        if beads["success"]:
            print(f"\nCreated epic: {beads.get('epic_id')}")
            print(f"Created {len(beads.get('phase_issues', []))} phase issues")

        # Step 6: Memory Capture (placeholder)
        print("\n" + "="*60)
        print("STEP 6/6: MEMORY CAPTURE")
        print("="*60)
        print("Memory capture: using existing hooks")
        results["steps"]["memory"] = {"success": True}

        # Complete - clean up any checkpoint for this run
        from .checkpoint_manager import detect_resumable_checkpoint, delete_checkpoint
        checkpoint = detect_resumable_checkpoint(self.project_path)
        if checkpoint:
            delete_checkpoint(checkpoint.get("file_path", ""))

        results["success"] = True
        results["completed"] = datetime.now().isoformat()
        results["plan_dir"] = str(Path(decomposition["phase_files"][0]).parent) if decomposition["phase_files"] else None
        results["epic_id"] = beads.get("epic_id")

        print("\n" + "="*60)
        print("PIPELINE COMPLETE")
        print("="*60)
        print(f"\nPlan directory: {results['plan_dir']}")
        print(f"Epic ID: {results['epic_id']}")

        return results

File: go/internal/planning/pipeline.go
package planning

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// PipelineConfig contains configuration for the planning pipeline.
type PipelineConfig struct {
	ProjectPath   string
	AutoApprove   bool
	TicketID      string
	ResearchPath  string       // Optional: path to existing research file (skips research step if provided)
	AutonomyMode  AutonomyMode
	MaxIterations int // Maximum iterations for implementation phase (default: IMPL_MAX_ITERATIONS)
}

// GetAutoApprove returns the auto-approve setting based on the autonomy mode.
func (pc *PipelineConfig) GetAutoApprove() bool {
	switch pc.AutonomyMode {
	case AutonomyCheckpoint:
		return false
	case AutonomyFullyAutonomous:
		return true
	case AutonomyBatch:
		return false // False at boundaries, true within groups (handled by orchestrator)
	default:
		return false
	}
}

// PipelineResults contains the results from all pipeline steps.
type PipelineResults struct {
	Success   bool                   `json:"success"`
	Started   string                 `json:"started"`
	Completed string                 `json:"completed,omitempty"`
	TicketID  string                 `json:"ticket_id,omitempty"`
	FailedAt  string                 `json:"failed_at,omitempty"`
	StoppedAt string                 `json:"stopped_at,omitempty"`
	Error     string                 `json:"error,omitempty"`
	PlanDir   string                 `json:"plan_dir,omitempty"`
	EpicID    string                 `json:"epic_id,omitempty"`
	Steps     map[string]interface{} `json:"steps"`
}

// PlanningPipeline orchestrates the 8-step planning and implementation process.
type PlanningPipeline struct {
	config PipelineConfig
}

// NewPlanningPipeline creates a new pipeline instance.
func NewPlanningPipeline(config PipelineConfig) *PlanningPipeline {
	return &PlanningPipeline{config: config}
}

// Run executes the complete planning pipeline.
func (p *PlanningPipeline) Run(researchPrompt string) *PipelineResults {
	results := &PipelineResults{
		Success:  true,
		Started:  time.Now().Format(time.RFC3339),
		TicketID: p.config.TicketID,
		Steps:    make(map[string]interface{}),
	}

	// Step 1: Research (skip if research path is provided)
	var research *StepResult
	if p.config.ResearchPath != "" {
		fmt.Println("\n" + strings.Repeat("=", 60))
		fmt.Println("STEP 1/8: RESEARCH PHASE (SKIPPED - using provided research)")
		fmt.Println(strings.Repeat("=", 60))
		fmt.Printf("Using existing research: %s\n", p.config.ResearchPath)

		// Create a synthetic result with the provided path
		research = NewStepResult()
		research.ResearchPath = p.config.ResearchPath
		research.Success = true
		results.Steps["research"] = research
	} else {
		fmt.Println("\n" + strings.Repeat("=", 60))
		fmt.Println("STEP 1/8: RESEARCH PHASE")
		fmt.Println(strings.Repeat("=", 60))

		research = StepResearch(p.config.ProjectPath, researchPrompt)
		results.Steps["research"] = research

		if !research.Success {
			results.Success = false
			results.FailedAt = "research"
			results.Error = research.Error
			return results
		}
	}

	// Step 2: Memory Sync
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 2/8: MEMORY SYNC")
	fmt.Println(strings.Repeat("=", 60))

	sessionID := fmt.Sprintf("research-%s", time.Now().Format("20060102-150405"))
	memoryResult := StepMemorySync(p.config.ProjectPath, research.ResearchPath, sessionID)
	results.Steps["memory_sync"] = memoryResult

	if episodeRecorded, ok := memoryResult.Data["episode_recorded"].(bool); ok && episodeRecorded {
		fmt.Println("  ✓ Episodic memory recorded")
	}
	if contextCompiled, ok := memoryResult.Data["context_compiled"].(bool); ok && contextCompiled {
		fmt.Println("  ✓ Working context compiled")
	}
	if contextCleared, ok := memoryResult.Data["context_cleared"].(bool); ok && contextCleared {
		fmt.Println("  ✓ Claude context cleared")
	}

	// Step 3: Requirement Decomposition
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 3/8: REQUIREMENT DECOMPOSITION")
	fmt.Println(strings.Repeat("=", 60))

	reqDecomp := StepRequirementDecomposition(p.config.ProjectPath, research.ResearchPath)
	results.Steps["requirement_decomposition"] = reqDecomp

	if reqDecomp.Success {
		fmt.Printf("\nDecomposed into %d requirements\n", reqDecomp.RequirementCount)
		fmt.Printf("Hierarchy: %s\n", reqDecomp.HierarchyPath)
	} else {
		fmt.Printf("\nDecomposition failed: %s\n", reqDecomp.Error)
		fmt.Println("Continuing to planning...")
	}

	// Step 4: Context Generation
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 4/8: CONTEXT GENERATION")
	fmt.Println(strings.Repeat("=", 60))

	contextGen := StepContextGeneration(p.config.ProjectPath, 100)
	results.Steps["context_generation"] = contextGen

	if contextGen.Success {
		fmt.Println("  ✓ Context generated successfully")
	} else {
		fmt.Printf("  ⚠ Context generation failed: %s\n", contextGen.Error)
		fmt.Println("  → Continuing without context (non-blocking)")
	}

	// Step 5: Planning
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 5/8: PLANNING PHASE")
	fmt.Println(strings.Repeat("=", 60))

	planning := StepPlanning(p.config.ProjectPath, research.ResearchPath, "")
	results.Steps["planning"] = planning

	if !planning.Success {
		results.Success = false
		results.FailedAt = "planning"
		results.Error = planning.Error
		return results
	}

	if planning.PlanPath == "" {
		results.Success = false
		results.FailedAt = "phase_decomposition"
		results.Error = "No plan_path extracted from planning step"
		return results
	}

	// Step 6: Phase Decomposition
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 6/8: PHASE DECOMPOSITION")
	fmt.Println(strings.Repeat("=", 60))

	decomposition := StepPhaseDecomposition(p.config.ProjectPath, planning.PlanPath)
	results.Steps["decomposition"] = decomposition

	if !decomposition.Success {
		results.Success = false
		results.FailedAt = "decomposition"
		results.Error = decomposition.Error
		return results
	}

	fmt.Printf("\nCreated %d phase files\n", len(decomposition.PhaseFiles))

	// Step 7: Beads Integration
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 7/8: BEADS INTEGRATION")
	fmt.Println(strings.Repeat("=", 60))

	epicTitle := fmt.Sprintf("Plan: %s", p.config.TicketID)
	if p.config.TicketID == "" {
		epicTitle = fmt.Sprintf("Plan: %s", time.Now().Format("2006-01-02"))
	}

	beads := StepBeadsIntegration(p.config.ProjectPath, decomposition.PhaseFiles, epicTitle)
	results.Steps["beads"] = beads

	if beads.Success {
		fmt.Printf("\nCreated epic: %s\n", beads.EpicID)
		fmt.Printf("Created %d phase issues\n", len(beads.PhaseIssues))
	}

	// Step 8: Implementation Phase
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 8/8: IMPLEMENTATION PHASE")
	fmt.Println(strings.Repeat("=", 60))

	// Extract issue IDs from phase issues
	issueIDs := getIssueIDsFromBeads(beads.PhaseIssues)

	// Get max iterations from config or use default
	maxIterations := p.config.MaxIterations
	if maxIterations == 0 {
		maxIterations = IMPL_MAX_ITERATIONS
	}

	// Run implementation phase
	impl := StepImplementation(
		p.config.ProjectPath,
		decomposition.PhaseFiles,
		issueIDs,
		beads.EpicID,
		maxIterations,
	)
	results.Steps["implementation"] = impl

	// Handle implementation result
	if !impl.Success {
		results.Success = false
		results.FailedAt = "implementation"
		results.Error = impl.Error

		// Add detailed error information
		fmt.Printf("\n❌ Implementation failed after %d iterations\n", impl.Iterations)
		if impl.Error != "" {
			fmt.Printf("Error: %s\n", impl.Error)
		}

		// Show which phases were completed
		if len(impl.PhasesClosed) > 0 {
			fmt.Printf("\nPhases completed: %v\n", impl.PhasesClosed)
		}

		// Show test status
		if !impl.TestsPassed && len(impl.PhasesClosed) == len(issueIDs) {
			fmt.Println("\n⚠ All phases closed but tests failed")
			if impl.Output != "" {
				// Show first 500 chars of test output
				testOutput := impl.Output
				if len(testOutput) > 500 {
					testOutput = testOutput[:500] + "...\n(truncated)"
				}
				fmt.Printf("\nTest output:\n%s\n", testOutput)
			}
		}

		fmt.Println("\n💡 Run 'bd ready' to see remaining work")

		results.Completed = time.Now().Format(time.RFC3339)
		if len(decomposition.PhaseFiles) > 0 {
			results.PlanDir = filepath.Dir(decomposition.PhaseFiles[0])
		}
		results.EpicID = beads.EpicID

		return results
	}

	// Success!
	fmt.Printf("\n✅ Implementation complete after %d iterations\n", impl.Iterations)
	fmt.Printf("✅ All %d phases closed\n", len(impl.PhasesClosed))
	fmt.Println("✅ All tests passed")

	// Complete
	results.Completed = time.Now().Format(time.RFC3339)
	if len(decomposition.PhaseFiles) > 0 {
		results.PlanDir = filepath.Dir(decomposition.PhaseFiles[0])
	}
	results.EpicID = beads.EpicID

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("PIPELINE COMPLETE")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("\nPlan directory: %s\n", results.PlanDir)
	fmt.Printf("Epic ID: %s\n", results.EpicID)

	return results
}

// getIssueIDsFromBeads extracts issue IDs from phase issues.
func getIssueIDsFromBeads(phaseIssues []PhaseIssue) []string {
	issueIDs := make([]string, 0, len(phaseIssues))
	for _, pi := range phaseIssues {
		if pi.IssueID != "" {
			issueIDs = append(issueIDs, pi.IssueID)
		}
	}
	return issueIDs
}

// ResumeFromCheckpoint resumes pipeline execution from a checkpoint.
func (p *PlanningPipeline) ResumeFromCheckpoint(checkpoint *Checkpoint, resumePhase string, planPath string) *PipelineResults {
	results := &PipelineResults{
		Success:  true,
		Started:  time.Now().Format(time.RFC3339),
		TicketID: p.config.TicketID,
		Steps:    make(map[string]interface{}),
	}

	fmt.Printf("\nResuming pipeline from %s phase...\n", resumePhase)

	// Determine which steps to run based on resume phase
	var research *StepResult
	var planning *StepResult

	switch resumePhase {
	case "research":
		// Resume from research step (run all steps)
		return p.Run("")

	case "planning":
		// Skip research, run from planning onwards
		if p.config.ResearchPath == "" {
			results.Success = false
			results.FailedAt = "planning"
			results.Error = "No research path provided for planning resume"
			return results
		}

		// Create synthetic research result
		research = NewStepResult()
		research.ResearchPath = p.config.ResearchPath
		research.Success = true
		results.Steps["research"] = research

		// Continue with memory sync and subsequent steps
		return p.runFromMemorySync(results, research)

	case "decomposition":
		// Skip to phase decomposition
		if planPath == "" {
			results.Success = false
			results.FailedAt = "decomposition"
			results.Error = "No plan path provided for decomposition resume"
			return results
		}

		// Create synthetic research result
		if p.config.ResearchPath != "" {
			research = NewStepResult()
			research.ResearchPath = p.config.ResearchPath
			research.Success = true
			results.Steps["research"] = research
		}

		// Create synthetic planning result
		planning = NewStepResult()
		planning.PlanPath = planPath
		planning.Success = true
		results.Steps["planning"] = planning

		// Run from phase decomposition
		return p.runFromPhaseDecomposition(results, planning)

	case "implementation":
		// Resume implementation phase
		// This requires phase files and beads info from checkpoint
		fmt.Println("⚠ Implementation phase resume not yet fully implemented")
		results.Success = false
		results.FailedAt = "implementation"
		results.Error = "Implementation phase resume requires phase files from checkpoint"
		return results

	default:
		results.Success = false
		results.FailedAt = resumePhase
		results.Error = fmt.Sprintf("Unknown resume phase: %s", resumePhase)
		return results
	}
}

// runFromMemorySync runs pipeline from memory sync step onwards.
func (p *PlanningPipeline) runFromMemorySync(results *PipelineResults, research *StepResult) *PipelineResults {
	// Step 2: Memory Sync
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 2/8: MEMORY SYNC")
	fmt.Println(strings.Repeat("=", 60))

	sessionID := fmt.Sprintf("research-%s", time.Now().Format("20060102-150405"))
	memoryResult := StepMemorySync(p.config.ProjectPath, research.ResearchPath, sessionID)
	results.Steps["memory_sync"] = memoryResult

	if episodeRecorded, ok := memoryResult.Data["episode_recorded"].(bool); ok && episodeRecorded {
		fmt.Println("  ✓ Episodic memory recorded")
	}
	if contextCompiled, ok := memoryResult.Data["context_compiled"].(bool); ok && contextCompiled {
		fmt.Println("  ✓ Working context compiled")
	}
	if contextCleared, ok := memoryResult.Data["context_cleared"].(bool); ok && contextCleared {
		fmt.Println("  ✓ Claude context cleared")
	}

	// Step 3: Requirement Decomposition
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 3/8: REQUIREMENT DECOMPOSITION")
	fmt.Println(strings.Repeat("=", 60))

	reqDecomp := StepRequirementDecomposition(p.config.ProjectPath, research.ResearchPath)
	results.Steps["requirement_decomposition"] = reqDecomp

	if reqDecomp.Success {
		fmt.Printf("\nDecomposed into %d requirements\n", reqDecomp.RequirementCount)
		fmt.Printf("Hierarchy: %s\n", reqDecomp.HierarchyPath)
	} else {
		fmt.Printf("\nDecomposition failed: %s\n", reqDecomp.Error)
		fmt.Println("Continuing to planning...")
	}

	// Step 4: Context Generation
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 4/8: CONTEXT GENERATION")
	fmt.Println(strings.Repeat("=", 60))

	contextGen := StepContextGeneration(p.config.ProjectPath, 100)
	results.Steps["context_generation"] = contextGen

	if contextGen.Success {
		fmt.Println("  ✓ Context generated successfully")
	} else {
		fmt.Printf("  ⚠ Context generation failed: %s\n", contextGen.Error)
		fmt.Println("  → Continuing without context (non-blocking)")
	}

	// Step 5: Planning
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 5/8: PLANNING PHASE")
	fmt.Println(strings.Repeat("=", 60))

	planning := StepPlanning(p.config.ProjectPath, research.ResearchPath, "")
	results.Steps["planning"] = planning

	if !planning.Success {
		results.Success = false
		results.FailedAt = "planning"
		results.Error = planning.Error
		return results
	}

	if planning.PlanPath == "" {
		results.Success = false
		results.FailedAt = "phase_decomposition"
		results.Error = "No plan_path extracted from planning step"
		return results
	}

	// Continue with phase decomposition
	return p.runFromPhaseDecomposition(results, planning)
}

// runFromPhaseDecomposition runs pipeline from phase decomposition onwards.
func (p *PlanningPipeline) runFromPhaseDecomposition(results *PipelineResults, planning *StepResult) *PipelineResults {
	// Step 6: Phase Decomposition
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 6/8: PHASE DECOMPOSITION")
	fmt.Println(strings.Repeat("=", 60))

	decomposition := StepPhaseDecomposition(p.config.ProjectPath, planning.PlanPath)
	results.Steps["decomposition"] = decomposition

	if !decomposition.Success {
		results.Success = false
		results.FailedAt = "decomposition"
		results.Error = decomposition.Error
		return results
	}

	fmt.Printf("\nCreated %d phase files\n", len(decomposition.PhaseFiles))

	// Step 7: Beads Integration
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 7/8: BEADS INTEGRATION")
	fmt.Println(strings.Repeat("=", 60))

	epicTitle := fmt.Sprintf("Plan: %s", p.config.TicketID)
	if p.config.TicketID == "" {
		epicTitle = fmt.Sprintf("Plan: %s", time.Now().Format("2006-01-02"))
	}

	beads := StepBeadsIntegration(p.config.ProjectPath, decomposition.PhaseFiles, epicTitle)
	results.Steps["beads"] = beads

	if beads.Success {
		fmt.Printf("\nCreated epic: %s\n", beads.EpicID)
		fmt.Printf("Created %d phase issues\n", len(beads.PhaseIssues))
	}

	// Step 8: Implementation Phase
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("STEP 8/8: IMPLEMENTATION PHASE")
	fmt.Println(strings.Repeat("=", 60))

	// Extract issue IDs from phase issues
	issueIDs := getIssueIDsFromBeads(beads.PhaseIssues)

	// Get max iterations from config or use default
	maxIterations := p.config.MaxIterations
	if maxIterations == 0 {
		maxIterations = IMPL_MAX_ITERATIONS
	}

	// Run implementation phase
	impl := StepImplementation(
		p.config.ProjectPath,
		decomposition.PhaseFiles,
		issueIDs,
		beads.EpicID,
		maxIterations,
	)
	results.Steps["implementation"] = impl

	// Handle implementation result
	if !impl.Success {
		results.Success = false
		results.FailedAt = "implementation"
		results.Error = impl.Error

		// Add detailed error information
		fmt.Printf("\n❌ Implementation failed after %d iterations\n", impl.Iterations)
		if impl.Error != "" {
			fmt.Printf("Error: %s\n", impl.Error)
		}

		// Show which phases were completed
		if len(impl.PhasesClosed) > 0 {
			fmt.Printf("\nPhases completed: %v\n", impl.PhasesClosed)
		}

		// Show test status
		if !impl.TestsPassed && len(impl.PhasesClosed) == len(issueIDs) {
			fmt.Println("\n⚠ All phases closed but tests failed")
			if impl.Output != "" {
				// Show first 500 chars of test output
				testOutput := impl.Output
				if len(testOutput) > 500 {
					testOutput = testOutput[:500] + "...\n(truncated)"
				}
				fmt.Printf("\nTest output:\n%s\n", testOutput)
			}
		}

		fmt.Println("\n💡 Run 'bd ready' to see remaining work")

		results.Completed = time.Now().Format(time.RFC3339)
		if len(decomposition.PhaseFiles) > 0 {
			results.PlanDir = filepath.Dir(decomposition.PhaseFiles[0])
		}
		results.EpicID = beads.EpicID

		return results
	}

	// Success!
	fmt.Printf("\n✅ Implementation complete after %d iterations\n", impl.Iterations)
	fmt.Printf("✅ All %d phases closed\n", len(impl.PhasesClosed))
	fmt.Println("✅ All tests passed")

	// Complete
	results.Completed = time.Now().Format(time.RFC3339)
	if len(decomposition.PhaseFiles) > 0 {
		results.PlanDir = filepath.Dir(decomposition.PhaseFiles[0])
	}
	results.EpicID = beads.EpicID

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("PIPELINE COMPLETE")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("\nPlan directory: %s\n", results.PlanDir)
	fmt.Printf("Epic ID: %s\n", results.EpicID)

	return results
}

// RequirementDecompositionResult contains results from requirement decomposition.
type RequirementDecompositionResult struct {
	Success          bool   `json:"success"`
	Error            string `json:"error,omitempty"`
	RequirementCount int    `json:"requirement_count"`
	HierarchyPath    string `json:"hierarchy_path,omitempty"`
	DiagramPath      string `json:"diagram_path,omitempty"`
}

// StepRequirementDecomposition decomposes research into structured requirements.
func StepRequirementDecomposition(projectPath, researchPath string) *RequirementDecompositionResult {
	result := &RequirementDecompositionResult{Success: true}

	// Read research content
	fullPath := researchPath
	if !filepath.IsAbs(researchPath) {
		fullPath = filepath.Join(projectPath, researchPath)
	}

	content, err := ReadFileContent(fullPath)
	if err != nil {
		result.Success = false
		result.Error = fmt.Sprintf("failed to read research: %v", err)
		return result
	}

	// Use decomposition function
	hierarchy, decompositionErr := DecomposeRequirements(content, projectPath, nil, nil, nil)
	if decompositionErr != nil {
		result.Success = false
		result.Error = decompositionErr.Message
		return result
	}

	// Count requirements
	result.RequirementCount = len(hierarchy.Requirements)
	for _, req := range hierarchy.Requirements {
		result.RequirementCount += len(req.Children)
	}

	// Save hierarchy to file
	dateStr := time.Now().Format("2006-01-02")
	outputDir := filepath.Join(projectPath, "thoughts", "searchable", "shared", "requirements")
	hierarchyPath := filepath.Join(outputDir, fmt.Sprintf("%s-hierarchy.json", dateStr))

	if err := SaveHierarchy(hierarchy, hierarchyPath); err != nil {
		result.Success = false
		result.Error = fmt.Sprintf("failed to save hierarchy: %v", err)
		return result
	}

	result.HierarchyPath = hierarchyPath
	return result
}

// ContextGenerationResult contains results from context generation.
type ContextGenerationResult struct {
	Success   bool   `json:"success"`
	Error     string `json:"error,omitempty"`
	OutputDir string `json:"output_dir,omitempty"`
}

// StepContextGeneration generates tech stack and file group context.
func StepContextGeneration(projectPath string, maxFiles int) *ContextGenerationResult {
	result := &ContextGenerationResult{Success: true}

	// This step analyzes the project structure and generates context
	// For now, we'll implement a basic version that scans the project

	dateStr := time.Now().Format("2006-01-02")
	outputDir := filepath.Join(projectPath, "thoughts", "searchable", "shared", "context", dateStr)

	// Create output directory
	if err := CreateDir(outputDir); err != nil {
		result.Success = false
		result.Error = fmt.Sprintf("failed to create output directory: %v", err)
		return result
	}

	result.OutputDir = outputDir
	return result
}

// ReadFileContent reads a file and returns its content as string.
func ReadFileContent(path string) (string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

// CreateDir creates a directory and all parent directories.
func CreateDir(path string) error {
	return os.MkdirAll(path, 0755)
}

// SaveHierarchy saves a requirement hierarchy to a JSON file.
func SaveHierarchy(hierarchy *RequirementHierarchy, path string) error {
	// Ensure parent directory exists
	if err := CreateDir(filepath.Dir(path)); err != nil {
		return err
	}

	data, err := hierarchy.ToJSON()
	if err != nil {
		return err
	}

	return os.WriteFile(path, data, 0644)
}
